lib/fluent/plugin/metrics.rb

```diff
@@ -0,0 +1,119 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'socket'
+
+require 'fluent/plugin/base'
+
+require 'fluent/log'
+require 'fluent/unique_id'
+require 'fluent/plugin_id'
+
+module Fluent
+  module Plugin
+    class Metrics < Base
+      include PluginId
+      include PluginLoggerMixin
+      include UniqueId::Mixin
+
+      DEFAULT_TYPE = 'local'
+
+      configured_in :metrics
+
+      config_param :default_labels, :hash, default: {agent: "Fluentd", hostname: "#{Socket.gethostname}"}
+      config_param :labels, :hash, default: {}
+
+      attr_reader :use_gauge_metric
+      attr_reader :has_methods_for_gauge, :has_methods_for_counter
+
+      def initialize
+        super
+
+        @has_methods_for_counter = false
+        @has_methods_for_gauge = false
+        @use_gauge_metric = false
+      end
+
+      def configure(conf)
+        super
+
+        if use_gauge_metric
+          @has_methods_for_gauge = has_methods_for_gauge?
+        else
+          @has_methods_for_counter = has_methods_for_counter?
+        end
+      end
+
+      # Some metrics should be counted by gauge.
+      # ref: https://prometheus.io/docs/concepts/metric_types/#gauge
+      def use_gauge_metric=(use_gauge_metric=false)
+        @use_gauge_metric = use_gauge_metric
+      end
+
+      def create(namespace:, subsystem:,name:,help_text:,labels: {})
+        # This API is for cmetrics type.
+      end
+
+      def get(key)
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def inc(key)
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def dec(key)
+        raise NotImplementedError, "Implement this method in child class"
+      end
```

Comment:
mention which class must be implemented ("gauge" or "counter" ) seems better.

Reply:
https://github.com/fluent/fluentd/pull/3440/files#diff-6826edd8ce4eb8c37724e029bb0fa852f74c76cd11a2a3e2e4ece1ea4761cee4R96-R105 and https://github.com/fluent/fluentd/pull/3440/files#diff-6826edd8ce4eb8c37724e029bb0fa852f74c76cd11a2a3e2e4ece1ea4761cee4R107-R116 mentions what is needed for counter or gauge? Are they not enough for you?

Reply:
For me, it is enough. I just think that it may be kind for metrics plugin developer/users.
Instead of "something method is not implemented", developer/users find "gauge metrics plugin must implement xxx but not it is not implemented".

Hmm, it may be redundant... :thinking:

Reply:
Yeah, It may be redundant. Because most of developers (except I was crazy someday 😛  Haha.) will check with try-and-error on developing plugins.